src/security/intel/cbnt/measurement.c

```diff
@@ -0,0 +1,543 @@
+	}
+
+	cbfs_unmap(bpm);
+	return CB_SUCCESS;
+}
+
+/*
+ * Pseudo-code of the data to be measured into PCR-7 for TigerLake and newer (older hardware
+ * isn't supported):
+ *
+ *     struct {
+ *         uint64_t ACM_POLICY_STATUS;
+ *         uint16_t ACM.Header.SVN;
+ *         uint8_t  ACM.KeyHash[32];
+ *         uint8_t  BP.KEY[32];
+ *         uint8_t  KM.BPKey.hashBuffer[BPM pubkey digest size];
+ *     } PCR7_DATA
+ *
+ * Applies only for TPM2.0.
+ *
+ * Returns true and initializes *hash on success.
+ */
+static bool make_pcr7_hash(struct obuf *ob, struct vb2_hash *hash)
+{
+	/*
+	 * ACM.KeyHash[32]
+	 *
+	 * BWG says to use TXT.PUBLIC.KEY, but TXT SDG (#315168-017.4) says that it
+	 * doesn't contain anything useful on modern platforms and suggests SGX-related
+	 * MSRs (0x20::0x23, to be specific, but SDM vol. 4 has these instead).
+	 */
+	(void)obuf_write_le64(ob, rdmsr(MSR_IA32_SGXLEPUBKEYHASH0).raw);
+	(void)obuf_write_le64(ob, rdmsr(MSR_IA32_SGXLEPUBKEYHASH1).raw);
+	(void)obuf_write_le64(ob, rdmsr(MSR_IA32_SGXLEPUBKEYHASH2).raw);
+	(void)obuf_write_le64(ob, rdmsr(MSR_IA32_SGXLEPUBKEYHASH3).raw);
+
+	/* BP.KEY[32] (rather its SHA256 hash) */
+	uint8_t bp_key[32];
+	size_t sz = sizeof(bp_key);
+	if (!read_cse_file(OEM_CRED_FILE, bp_key, &sz, 0, MKHI_READ_FILE_FLAG_DEFAULT)) {
+		printk(BIOS_ERR,
+		       "CBnT: failed to obtain BP.KEY hash for PCR-7 measurement data\n");
+		return false;
+	}
+	if (obuf_write(ob, bp_key, sizeof(bp_key))) {
+		printk(BIOS_ERR, "CBnT: failed to write ACM's key hash\n");
+		return false;
+	}
+
+	if (fill_pcr7_km_fields(ob) != CB_SUCCESS) {
+		printk(BIOS_ERR, "CBnT: failed to fill KM fields of PCR-7 measurement data\n");
+		return false;
+	}
+
+	size_t size;
+	const void *data = obuf_contents(ob, &size);
+	if (vb2_hash_calculate(vboot_hwcrypto_allowed(), data, size, tpm_log_alg(), hash)) {
+		printk(BIOS_ERR, "CBnT: failed to hash PCR-7 measurement data\n");
+		return false;
+	}
+
+	return true;
+}
+
+void intel_cbnt_inject_ibg_measurements(void)
+{
+	const union cbnt_biosacm_policy biosacm_sts = {
+		.raw = read64p(CBNT_BIOSACM_POLICY_STS),
+	};
+
+	/* Do nothing if BootGuard wasn't involved in the boot process. */
+	if (biosacm_sts.status.scrtm_status == 0)
+		return;
+
+	/* Do nothing if BootGuard doesn't measure anything. */
+	if (!(biosacm_sts.status.bp & CBNT_BP_TYPE_M))
+		return;
+
+	/*
+	 * At this point the code should check whether PCRs 0-7 were capped with a digest of
+	 * value "1" (ASCII? binary? hex?) and if so, add appropriate entries to the event log.
+	 *
+	 * This can only happen if Bit3 of IbbFlags is set ("On error: Leave TPM Hierarchies
+	 * enabled. Cap all PCRs"), otherwise TPM is just disabled.  Either way Measured Boot
+	 * won't work, so skipping it unless reasons for doing it are discovered.
+	 */
+
+	/*
+	 * Making and hashing PCR-7 data.
+	 *
+	 * Pseudo-code of the data to be measured into PCR-0 for TigerLake and newer (older
+	 * hardware isn't supported):
```

Comment:
```suggestion
	 * hardware isn't supported yet):
```

But the process is pretty much the same (just slightly different way of obtaining the data).

Reply:
Updated: https://github.com/Dasharo/coreboot/compare/ffa516f0285a87e049c9e7ebefc0b23e3d941a77..4b501619643dc15f6eb104ad64919f0fb0b5e347#diff-1725d4cd7234cc446584ac6ebad791ed8c8e640970d4adb2e99e5b5b233030f7R534